src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -4,274 +4,291 @@
-                if (tokens.Length == 2)
+                using (var server = new LoopbackServer(listenSocket, options))
                 {
-                    string headerName = tokens[0];
-                    if (headerName == "Sec-WebSocket-Key")
-                    {
-                        string headerValue = tokens[1].Trim();
-                        string responseSecurityAcceptValue = ComputeWebSocketHandshakeSecurityAcceptValue(headerValue);
-                        serverResponse =
-                            "HTTP/1.1 101 Switching Protocols\r\n" +
-                            "Upgrade: websocket\r\n" +
-                            "Connection: Upgrade\r\n" +
-                            "Sec-WebSocket-Accept: " + responseSecurityAcceptValue + "\r\n\r\n";
-                    }
+                    await funcAsync(server);
                 }
             }
+        }
 
-            if (serverResponse != null)
-            {
-                // We received a valid WebSocket opening handshake. Send the appropriate response.
-                await writer.WriteAsync(serverResponse).ConfigureAwait(false);
-                return true;
-            }
-
-            return false;
+        public static Task CreateServerAsync(Func<LoopbackServer, Uri, Task> funcAsync, Options options = null)
+        {
+            return CreateServerAsync(server => funcAsync(server, server.Uri), options);
         }
 
-        private static string ComputeWebSocketHandshakeSecurityAcceptValue(string secWebSocketKey)
+        public static Task CreateClientAndServerAsync(Func<Uri, Task> clientFunc, Func<LoopbackServer, Task> serverFunc)
         {
-            // GUID specified by RFC 6455.
-            const string Rfc6455Guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
-            string combinedKey = secWebSocketKey + Rfc6455Guid;
-
-            // Use of SHA1 hash is required by RFC 6455.
-            SHA1 sha1Provider = new SHA1CryptoServiceProvider();
-            byte[] sha1Hash = sha1Provider.ComputeHash(Encoding.UTF8.GetBytes(combinedKey));
-            return Convert.ToBase64String(sha1Hash);
+            return CreateServerAsync(async server =>
+            {
+                Task clientTask = clientFunc(server.Uri);
+                Task serverTask = serverFunc(server);
+
+                await new Task[] { clientTask, serverTask }.WhenAllOrAnyFailed();
+            });
         }
 
-        public static async Task<List<string>> AcceptSocketAsync(Socket server, Func<Socket, Stream, StreamReader, StreamWriter, Task<List<string>>> funcAsync, Options options = null)
+        public async Task AcceptConnectionAsync(Func<Connection, Task> funcAsync)
         {
-            options = options ?? new Options();
-            Socket s = await server.AcceptAsync().ConfigureAwait(false);
-            s.NoDelay = true;
-            try
+            using (Socket s = await _listenSocket.AcceptAsync().ConfigureAwait(false))
```

Comment:
It shouldn't come to that but adding a timeout (larger than the default for the handlers) seems like a good idea just to avoid the test deadlocking here.

Reply:
This does seem like a good idea; I'm going to defer for now in the interest of getting this PR in, but we should consider for the future.